src/test/scala/edu/berkeley/cs/rise/opaque/AggregationSuite.scala

```diff
@@ -0,0 +1,492 @@
+          |  MAX(null),
+          |  MIN(null),
+          |  SUM(null)
+        """.stripMargin)
+    }
+  }
+
+  test("only do grouping") {
+    checkAnswer() { sl =>
+      loadAggData(sl)
+      spark.sql("""
+          |SELECT key
+          |FROM agg1
+          |GROUP BY key
+        """.stripMargin),
+    }
+
+    checkAnswer() { sl =>
+      loadAggData(sl)
+      spark.sql("""
+          |SELECT DISTINCT value1, key
+          |FROM agg2
+        """.stripMargin),
+    }
+
+    checkAnswer() { sl =>
+      loadAggData(sl)
+      spark.sql("""
+          |SELECT value1, key
+          |FROM agg2
+          |GROUP BY key, value1
+        """.stripMargin),
+    }
+
+    checkAnswer(ignore = true) { sl =>
+      loadAggData(sl)
+      spark.sql("""
+          |SELECT DISTINCT key
+          |FROM agg3
+        """.stripMargin),
+    }
+
+    checkAnswer(ignore = true) { sl =>
+      loadAggData(sl)
+      spark.sql("""
+          |SELECT value1, key
+          |FROM agg3
+          |GROUP BY value1, key
+        """.stripMargin)
+    }
+  }
+
+  ignore("case in-sensitive resolution") {
+    checkAnswer() { sl =>
+      loadAggData(sl)
+      spark.sql("""
+          |SELECT avg(value), kEY - 100
+          |FROM agg1
+          |GROUP BY Key - 100
+        """.stripMargin)
+    }
+
+    checkAnswer() { sl =>
+      loadAggData(sl)
+      spark.sql("""
+          |SELECT sum(distinct value1), kEY - 100, count(distinct value1)
+          |FROM agg2
+          |GROUP BY Key - 100
+        """.stripMargin)
+    }
+
+    checkAnswer() { sl =>
+      loadAggData(sl)
+      spark.sql("""
+          |SELECT valUe * key - 100
+          |FROM agg1
+          |GROUP BY vAlue * keY - 100
+        """.stripMargin)
+    }
+  }
+
+  test("test average no key in output") {
+    checkAnswer() { sl =>
+      loadAggData(sl)
+      spark.sql("""
+          |SELECT avg(value)
+          |FROM agg1
+          |GROUP BY key
+        """.stripMargin)
+    }
+  }
+
+  test("test average") {
+    checkAnswer() { sl =>
+      loadAggData(sl)
+      spark.sql("""
+          |SELECT key, avg(value)
+          |FROM agg1
+          |GROUP BY key
+        """.stripMargin)
+    }
+
+    checkAnswer() { sl =>
+      loadAggData(sl)
+      spark.sql("""
+          |SELECT key, mean(value)
+          |FROM agg1
+          |GROUP BY key
+        """.stripMargin)
+    }
+
+    checkAnswer() { sl =>
+      loadAggData(sl)
+      spark.sql("""
+          |SELECT avg(value), key
+          |FROM agg1
+          |GROUP BY key
+        """.stripMargin)
+    }
+
+    checkAnswer(true) { sl =>
+      loadAggData(sl)
+      spark.sql("""
+          |SELECT avg(value) + 1.5, key + 10
+          |FROM agg1
+          |GROUP BY key + 10
+        """.stripMargin)
+    }
+
+    checkAnswer() { sl =>
+      loadAggData(sl)
+      spark.sql("""
+          |SELECT avg(value) FROM agg1
+        """.stripMargin)
+    }
+  }
+
+  ignore("first_value and last_value") {
```

Comment:
Will we have support for this in the future?

Reply:
I would think so? Seems like a function that is supported by Spark SQL https://spark.apache.org/docs/3.1.1/api/sql/index.html#first_value